src/nvim/terminal.c

```diff
@@ -138,6 +138,8 @@ struct terminal {
   int pressed_button;               // which mouse button is pressed
   bool pending_resize;              // pending width/height
 
+  bool color_set[16];
```

Comment:
Could this be extended to 256 colors (or 21 for base16 based schemes)?
(I'm not really sure how/where this makes a difference though)
For me it looks ok as-is, and it appears to affect only being able to set it via the `b:/g:terminal_color_X` vars, doesn't it?

With the following patch I can set them, but they are off then (dark blue/black).  It seems like setting "green" to 16 is dark black, while it is a bit more blueish when setting it to 17 then:
```diff
 src/nvim/terminal.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git i/src/nvim/terminal.c w/src/nvim/terminal.c
index c5e756905..c78b536f0 100644
--- i/src/nvim/terminal.c
+++ w/src/nvim/terminal.c
@@ -138,7 +138,7 @@ struct terminal {
   int pressed_button;               // which mouse button is pressed
   bool pending_resize;              // pending width/height
 
-  bool color_set[16];
+  bool color_set[22];
 
   size_t refcount;                  // reference count
 };
@@ -227,7 +227,7 @@ Terminal *terminal_open(TerminalOptions opts)
   // - b:terminal_color_{NUM}
   // - g:terminal_color_{NUM}
   // - the VTerm instance
-  for (int i = 0; i < 16; i++) {
+  for (int i = 0; i < 22; i++) {
     RgbValue color_val = -1;
     char var[64];
     snprintf(var, sizeof(var), "terminal_color_%d", i);
@@ -607,8 +607,8 @@ void terminal_get_line_attributes(Terminal *term, win_T *wp, int linenr,
     int vt_fg_idx = ((!fg_default && fg_indexed) ? cell.fg.indexed.idx + 1 : 0);
     int vt_bg_idx = ((!bg_default && bg_indexed) ? cell.bg.indexed.idx + 1 : 0);
 
-    bool fg_set = vt_fg_idx && vt_fg_idx <= 16 && term->color_set[vt_fg_idx-1];
-    bool bg_set = vt_bg_idx && vt_bg_idx <= 16 && term->color_set[vt_bg_idx-1];
+    bool fg_set = vt_fg_idx && vt_fg_idx <= 22 && term->color_set[vt_fg_idx-1];
+    bool bg_set = vt_bg_idx && vt_bg_idx <= 22 && term->color_set[vt_bg_idx-1];
 
     int hl_attrs = (cell.attrs.bold ? HL_BOLD : 0)
                  | (cell.attrs.italic ? HL_ITALIC : 0)
```

Reply:
@blueyed,

I mention over in [#7018](https://github.com/neovim/neovim/issues/7018) that this PR would fix that issue.

Note, I **needed** to emphasise that it would be fixed if colours 16-255 were inherited from the terminal, especially if `termguicolors` is set.